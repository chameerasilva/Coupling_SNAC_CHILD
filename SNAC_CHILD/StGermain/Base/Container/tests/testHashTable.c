/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** $Id: testHashTable.c 2136 2005-02-28 02:47:13Z RaquibulHassan $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include "Base/Foundation/Foundation.h"
#include "Base/IO/IO.h"
#include "Base/Container/Container.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

void deleteFunction( void *data ){
	assert( data );

	Memory_Free( data );
}

void printFunction( void *data, void *stream ){
	
	assert( data );
	assert( stream );

	Journal_Printf( (Stream*)stream, "\t\t%s", (void*)data );
}

void ptrPrintFunction( void *data, void *stream ){
	
	assert( data );
	assert( stream );

	Journal_Printf( (Stream*)stream, "\t\t(ptr)%p\n", (void*)data );
}

int main( int argc, char* argv[] ) {
	MPI_Comm			CommWorld;
	int				rank;
	int				numProcessors;
	int				procToWatch;
	HashTable		*table;
	HashTable		*ptrTable;
	Stream *myStream = NULL;
	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_dup( MPI_COMM_WORLD, &CommWorld );
	MPI_Comm_size( CommWorld, &numProcessors );
	MPI_Comm_rank( CommWorld, &rank );

	BaseFoundation_Init( &argc, &argv );
	BaseIO_Init( &argc, &argv );
	BaseContainer_Init( &argc, &argv );

	if( argc >= 2 ) {
		procToWatch = atoi( argv[1] );
	}
	else {
		procToWatch = 0;
	}
	
	if( rank == procToWatch ) {
		FILE *dataFile = NULL;
		char buffer[512];
		int result = 0;
		int sum = 0;
		int count = 0;
		char *data = NULL;
		char *searchResult = NULL;
		void *dataRefs[200];
		
		myStream = Journal_Register( InfoStream_Type, "hashTableStream" );

		table = HashTable_New( NULL, printFunction, NULL, HASHTABLE_STRING_KEY );
		
		Journal_Printf( myStream, "Testing hash table with string keys..\n\n" );
		Journal_Printf( myStream, "Inserting data into the hash table\n\n" );
		
		dataFile = fopen( "dictionary.txt", "r+" );
		memset( buffer, 0, sizeof( buffer ) );
		while( (fgets( buffer, sizeof(buffer), dataFile ) != NULL) ){
			data = StG_Strdup(buffer);
			result = HashTable_InsertEntry( table, (void*)data, strlen(buffer), (void*)data, strlen(buffer));
			if( result > 1 ){
				Journal_Printf( myStream, "Incurred %d collisions while inserting %s", result, buffer );
			}
			sum += result;
			memset( buffer, 0, sizeof( buffer ) );
		}
		fclose( dataFile );
		
		Journal_Printf( myStream, "Successfully inserted %d items into the table\n\n", table->count );

		if( sum == table->count ){
			Journal_Printf( myStream, "0 collisions\n\n" );
		}
		
		Journal_Printf( myStream, "Printing the hash table\n\n" );
		Print( table, myStream );

		Journal_Printf( myStream, "Searching for items in the table:\n\n" );
		
		dataFile = fopen( "dictionary.txt", "r+" );
		memset( buffer, 0, sizeof( buffer ) );
		count = 0;
		while( (fgets( buffer, sizeof(buffer), dataFile ) != NULL) ){
			searchResult = (char*)HashTable_FindEntry( table, (void*)buffer, strlen(buffer), char* );
			assert( searchResult );
			if( !strcmp( buffer, searchResult ) ){
				searchResult[strlen(searchResult)-1] = '\0';
				Journal_Printf( myStream, "Looking for %s, ...found\n", searchResult );
			}
			else{
				Journal_Printf( myStream, "item not found..!!\n" );
			}
			memset( buffer, 0, sizeof( buffer ) );
			count++;
		}
		fclose( dataFile );
	
		Journal_Printf( myStream, "\nDeleting the Table\n\n" );
		Stg_Class_Delete( table );
		
		ptrTable = HashTable_New( NULL, ptrPrintFunction, NULL, HASHTABLE_POINTER_KEY );
		
		Journal_Printf( myStream, "Testing hash table with pointer keys..\n\n" );
		Journal_Printf( myStream, "Inserting data into the hash table\n\n" );
		
		dataFile = fopen( "dictionary.txt", "r+" );
		memset( buffer, 0, sizeof( buffer ) );
		count = 0;
		while( (fgets( buffer, sizeof(buffer), dataFile ) != NULL) ){
			data = StG_Strdup(buffer);
			result = HashTable_InsertEntry( ptrTable, (void*)data, sizeof(void*), (void*)data, strlen(buffer));
			if( result > 1 ){
				Journal_Printf( myStream, "Incurred %d collisions while inserting (ptr)%p\n", result, data );
			}
			sum += result;
			memset( buffer, 0, sizeof( buffer ) );
			dataRefs[count++] = data;
		}
		fclose( dataFile );
		
		Journal_Printf( myStream, "Successfully inserted %d items into the table\n\n", ptrTable->count );

		if( sum == ptrTable->count ){
			Journal_Printf( myStream, "0 collisions\n\n" );
		}
		
		Journal_Printf( myStream, "Printing the hash table\n\n" );
		Print( ptrTable, myStream );

		Journal_Printf( myStream, "Searching for items in the table:\n\n" );
		
		dataFile = fopen( "dictionary.txt", "r+" );
		memset( buffer, 0, sizeof( buffer ) );
		count = 0;
		while( (fgets( buffer, sizeof(buffer), dataFile ) != NULL) ){
			searchResult = (char*)HashTable_FindEntry( ptrTable, (void*)dataRefs[count], strlen(buffer), char* );
			assert( searchResult );
			if( !strcmp( buffer, searchResult ) ){
				searchResult[strlen(searchResult)-1] = '\0';
				Journal_Printf( myStream, "Looking for (ptr)%p, ...found\n", searchResult );
			}
			else{
				Journal_Printf( myStream, "item not found..!!\n" );
			}
			memset( buffer, 0, sizeof( buffer ) );
			count++;
		}
		fclose( dataFile );
	
		Journal_Printf( myStream, "\nDeleting the Table\n" );
		Stg_Class_Delete( ptrTable );
	}

	BaseContainer_Finalise();
	BaseIO_Finalise();
	BaseFoundation_Finalise();
	
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}
