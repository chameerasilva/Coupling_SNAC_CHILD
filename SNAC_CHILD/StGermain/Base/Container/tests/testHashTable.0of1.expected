Testing hash table with string keys..

Inserting data into the hash table

Incurred 2 collisions while inserting anyplace
Incurred 2 collisions while inserting electronic
Incurred 2 collisions while inserting drawl
Incurred 2 collisions while inserting buggy
Successfully inserted 100 items into the table

Printing the hash table

Stg_Class (ptr): 0x80cc3cc
	sizeOfSelf: 72
	_deleteSelf: Yes
	type: HashTable
	_delete (func ptr): 0x805c0a6
	_print (func ptr): 0x805bf61
	_copy (func ptr): (nil)
HashTable (ptr): (0x80cc3cc)
	EntryCount		 - 100
	HashTable KeyType	 - STRING
	HashTable data	 - 
		cabinet
		univariate
		antarctic
		anatomy
		exogamy
		escheat
		cardioid
		solidus
		embolden
		memoir
		piteous
		felt
		frown
		fanout
		kibitz
		anyplace
		cavort
		clipboard
		divorcee
		allowance
		melodic
		grillwork
		magnitude
		fiduciary
		up
		semester
		threat
		electronic
		plunge
		infantrymen
		wildlife
		head
		excrescent
		steepen
		sepal
		exhale
		insuppressible
		gumbo
		heal
		drawl
		presentation
		whinny
		textile
		slugging
		free
		involve
		controller
		orchestrate
		amoeba
		blurt
		eerie
		thanksgiving
		loin
		jiggle
		biology
		apace
		fairy
		pattern
		stance
		buggy
		circumspect
		buret
		sudden
		sequent
		monitor
		diabase
		ephemerides
		bandstand
		nitride
		arhat
		clothesmen
		conscript
		assassin
		firemen
		us
		depression
		bastard
		climatology
		nether
		posthumous
		virile
		caulk
		etude
		germinate
		licensor
		potlatch
		stanch
		crucifixion
		millions
		misshapen
		lottery
		worth
		chert
		observe
		fidelity
		precaution
		intelligentsia
		bleed
		kennel
		town
Searching for items in the table:

Looking for antarctic, ...found
Looking for escheat, ...found
Looking for cardioid, ...found
Looking for solidus, ...found
Looking for memoir, ...found
Looking for piteous, ...found
Looking for felt, ...found
Looking for frown, ...found
Looking for fanout, ...found
Looking for allowance, ...found
Looking for up, ...found
Looking for semester, ...found
Looking for threat, ...found
Looking for plunge, ...found
Looking for head, ...found
Looking for sepal, ...found
Looking for gumbo, ...found
Looking for textile, ...found
Looking for slugging, ...found
Looking for blurt, ...found
Looking for thanksgiving, ...found
Looking for apace, ...found
Looking for pattern, ...found
Looking for stance, ...found
Looking for circumspect, ...found
Looking for buret, ...found
Looking for sudden, ...found
Looking for sequent, ...found
Looking for monitor, ...found
Looking for diabase, ...found
Looking for bandstand, ...found
Looking for nitride, ...found
Looking for clothesmen, ...found
Looking for assassin, ...found
Looking for firemen, ...found
Looking for us, ...found
Looking for depression, ...found
Looking for bastard, ...found
Looking for climatology, ...found
Looking for nether, ...found
Looking for posthumous, ...found
Looking for virile, ...found
Looking for caulk, ...found
Looking for etude, ...found
Looking for germinate, ...found
Looking for licensor, ...found
Looking for potlatch, ...found
Looking for stanch, ...found
Looking for crucifixion, ...found
Looking for millions, ...found
Looking for misshapen, ...found
Looking for lottery, ...found
Looking for worth, ...found
Looking for chert, ...found
Looking for observe, ...found
Looking for fidelity, ...found
Looking for precaution, ...found
Looking for intelligentsia, ...found
Looking for bleed, ...found
Looking for kennel, ...found
Looking for town, ...found
Looking for cabinet, ...found
Looking for univariate, ...found
Looking for anatomy, ...found
Looking for exogamy, ...found
Looking for embolden, ...found
Looking for kibitz, ...found
Looking for anyplace, ...found
Looking for cavort, ...found
Looking for clipboard, ...found
Looking for divorcee, ...found
Looking for melodic, ...found
Looking for grillwork, ...found
Looking for magnitude, ...found
Looking for fiduciary, ...found
Looking for electronic, ...found
Looking for infantrymen, ...found
Looking for wildlife, ...found
Looking for excrescent, ...found
Looking for steepen, ...found
Looking for exhale, ...found
Looking for insuppressible, ...found
Looking for heal, ...found
Looking for drawl, ...found
Looking for presentation, ...found
Looking for whinny, ...found
Looking for free, ...found
Looking for involve, ...found
Looking for controller, ...found
Looking for orchestrate, ...found
Looking for amoeba, ...found
Looking for eerie, ...found
Looking for loin, ...found
Looking for jiggle, ...found
Looking for biology, ...found
Looking for fairy, ...found
Looking for buggy, ...found
Looking for ephemerides, ...found
Looking for arhat, ...found
Looking for conscript, ...found

Deleting the Table

Testing hash table with pointer keys..

Inserting data into the hash table

Successfully inserted 100 items into the table

Printing the hash table

Stg_Class (ptr): 0x80ce714
	sizeOfSelf: 72
	_deleteSelf: Yes
	type: HashTable
	_delete (func ptr): 0x805c0a6
	_print (func ptr): 0x805bf61
	_copy (func ptr): (nil)
HashTable (ptr): (0x80ce714)
	EntryCount		 - 100
	HashTable KeyType	 - INTEGER
	HashTable data	 - 
		(ptr)0x80cca78
		(ptr)0x80ce7e8
		(ptr)0x80cf8b0
		(ptr)0x80cfa48
		(ptr)0x80cfa78
		(ptr)0x80cfac8
		(ptr)0x80ccfa0
		(ptr)0x80cf678
		(ptr)0x80cfb70
		(ptr)0x80cd698
		(ptr)0x80cd708
		(ptr)0x80cddb0
		(ptr)0x80cf928
		(ptr)0x80cf6a8
		(ptr)0x80cdfb8
		(ptr)0x80cd128
		(ptr)0x80cfc18
		(ptr)0x80cd288
		(ptr)0x80cf9f8
		(ptr)0x80cd930
		(ptr)0x80ccae8
		(ptr)0x80cf650
		(ptr)0x80cf9d0
		(ptr)0x80cfbc8
		(ptr)0x80ccb58
		(ptr)0x80cf7e8
		(ptr)0x80ccca8
		(ptr)0x80cf8d8
		(ptr)0x80cf9a8
		(ptr)0x80cf798
		(ptr)0x80cdba0
		(ptr)0x80ccb90
		(ptr)0x80cfc68
		(ptr)0x80cd890
		(ptr)0x80cf888
		(ptr)0x80ce810
		(ptr)0x80cd198
		(ptr)0x80cd5a8
		(ptr)0x80cfa20
		(ptr)0x80cdaf0
		(ptr)0x80cdd40
		(ptr)0x80cd2f8
		(ptr)0x80cf810
		(ptr)0x80ce0a0
		(ptr)0x80cdc90
		(ptr)0x80cda48
		(ptr)0x80cfaa0
		(ptr)0x80cd3d8
		(ptr)0x80cd9d8
		(ptr)0x80cd538
		(ptr)0x80cdf10
		(ptr)0x80cca40
		(ptr)0x80cfbf0
		(ptr)0x80cce40
		(ptr)0x80cce78
		(ptr)0x80cf900
		(ptr)0x80cf7c0
		(ptr)0x80cd010
		(ptr)0x80cd778
		(ptr)0x80cc810
		(ptr)0x80cd4c8
		(ptr)0x80cfaf8
		(ptr)0x80ce7c0
		(ptr)0x80cf6d0
		(ptr)0x80cf950
		(ptr)0x80cfb48
		(ptr)0x80ccd88
		(ptr)0x80cf770
		(ptr)0x80cc800
		(ptr)0x80cf860
		(ptr)0x80cf748
		(ptr)0x80cd210
		(ptr)0x80cfcb8
		(ptr)0x80ccd18
		(ptr)0x80cf838
		(ptr)0x80ce068
		(ptr)0x80cf720
		(ptr)0x80cd7b0
		(ptr)0x80ccc00
		(ptr)0x80cd450
		(ptr)0x80cf6f8
		(ptr)0x80ce838
		(ptr)0x80cd368
		(ptr)0x80cfd10
		(ptr)0x80ccf30
		(ptr)0x80cfd38
		(ptr)0x80cde28
		(ptr)0x80cfc40
		(ptr)0x80cfce0
		(ptr)0x80cda80
		(ptr)0x80cfb98
		(ptr)0x80cc9d0
		(ptr)0x80cfb20
		(ptr)0x80cd618
		(ptr)0x80ccdc0
		(ptr)0x80cfc90
		(ptr)0x80cce00
		(ptr)0x80cf978
		(ptr)0x80cd048
		(ptr)0x80cdbd8
Searching for items in the table:

Looking for (ptr)0x80cd708, ...found
Looking for (ptr)0x80cd698, ...found
Looking for (ptr)0x80cd618, ...found
Looking for (ptr)0x80cd5a8, ...found
Looking for (ptr)0x80cd538, ...found
Looking for (ptr)0x80cd4c8, ...found
Looking for (ptr)0x80cd450, ...found
Looking for (ptr)0x80cd3d8, ...found
Looking for (ptr)0x80cd368, ...found
Looking for (ptr)0x80cd2f8, ...found
Looking for (ptr)0x80cd288, ...found
Looking for (ptr)0x80cd210, ...found
Looking for (ptr)0x80cd198, ...found
Looking for (ptr)0x80cd128, ...found
Looking for (ptr)0x80ce0a0, ...found
Looking for (ptr)0x80ce068, ...found
Looking for (ptr)0x80cd048, ...found
Looking for (ptr)0x80cd010, ...found
Looking for (ptr)0x80ccfa0, ...found
Looking for (ptr)0x80ccf30, ...found
Looking for (ptr)0x80cdfb8, ...found
Looking for (ptr)0x80cce78, ...found
Looking for (ptr)0x80cce40, ...found
Looking for (ptr)0x80cdf10, ...found
Looking for (ptr)0x80cce00, ...found
Looking for (ptr)0x80ccdc0, ...found
Looking for (ptr)0x80cde28, ...found
Looking for (ptr)0x80cddb0, ...found
Looking for (ptr)0x80ccd88, ...found
Looking for (ptr)0x80cdd40, ...found
Looking for (ptr)0x80cdc90, ...found
Looking for (ptr)0x80ccd18, ...found
Looking for (ptr)0x80cdbd8, ...found
Looking for (ptr)0x80cdba0, ...found
Looking for (ptr)0x80ccca8, ...found
Looking for (ptr)0x80cdaf0, ...found
Looking for (ptr)0x80cda80, ...found
Looking for (ptr)0x80ccc00, ...found
Looking for (ptr)0x80cda48, ...found
Looking for (ptr)0x80cd9d8, ...found
Looking for (ptr)0x80ccb90, ...found
Looking for (ptr)0x80cd930, ...found
Looking for (ptr)0x80cd890, ...found
Looking for (ptr)0x80ccb58, ...found
Looking for (ptr)0x80ccae8, ...found
Looking for (ptr)0x80cca78, ...found
Looking for (ptr)0x80cca40, ...found
Looking for (ptr)0x80cc9d0, ...found
Looking for (ptr)0x80cd7b0, ...found
Looking for (ptr)0x80cd778, ...found
Looking for (ptr)0x80cc800, ...found
Looking for (ptr)0x80cc810, ...found
Looking for (ptr)0x80ce7c0, ...found
Looking for (ptr)0x80ce7e8, ...found
Looking for (ptr)0x80ce810, ...found
Looking for (ptr)0x80ce838, ...found
Looking for (ptr)0x80cf650, ...found
Looking for (ptr)0x80cf678, ...found
Looking for (ptr)0x80cf6a8, ...found
Looking for (ptr)0x80cf6d0, ...found
Looking for (ptr)0x80cf6f8, ...found
Looking for (ptr)0x80cf720, ...found
Looking for (ptr)0x80cf748, ...found
Looking for (ptr)0x80cf770, ...found
Looking for (ptr)0x80cf798, ...found
Looking for (ptr)0x80cf7c0, ...found
Looking for (ptr)0x80cf7e8, ...found
Looking for (ptr)0x80cf810, ...found
Looking for (ptr)0x80cf838, ...found
Looking for (ptr)0x80cf860, ...found
Looking for (ptr)0x80cf888, ...found
Looking for (ptr)0x80cf8b0, ...found
Looking for (ptr)0x80cf8d8, ...found
Looking for (ptr)0x80cf900, ...found
Looking for (ptr)0x80cf928, ...found
Looking for (ptr)0x80cf950, ...found
Looking for (ptr)0x80cf978, ...found
Looking for (ptr)0x80cf9a8, ...found
Looking for (ptr)0x80cf9d0, ...found
Looking for (ptr)0x80cf9f8, ...found
Looking for (ptr)0x80cfa20, ...found
Looking for (ptr)0x80cfa48, ...found
Looking for (ptr)0x80cfa78, ...found
Looking for (ptr)0x80cfaa0, ...found
Looking for (ptr)0x80cfac8, ...found
Looking for (ptr)0x80cfaf8, ...found
Looking for (ptr)0x80cfb20, ...found
Looking for (ptr)0x80cfb48, ...found
Looking for (ptr)0x80cfb70, ...found
Looking for (ptr)0x80cfb98, ...found
Looking for (ptr)0x80cfbc8, ...found
Looking for (ptr)0x80cfbf0, ...found
Looking for (ptr)0x80cfc18, ...found
Looking for (ptr)0x80cfc40, ...found
Looking for (ptr)0x80cfc68, ...found
Looking for (ptr)0x80cfc90, ...found
Looking for (ptr)0x80cfcb8, ...found
Looking for (ptr)0x80cfce0, ...found
Looking for (ptr)0x80cfd10, ...found
Looking for (ptr)0x80cfd38, ...found

Deleting the Table
